File: tutorials/mpi/examples/ser_prime.c
/******************************************************************************
* FILE: ser_prime.c
* DESCRIPTION: 
*   This program generates primes. The approach taken is a "brute force"
*   method which requires increasingly greater amounts of cpu as the problem
*   size increases. It should lend itself well to an embarrassingly parallel
*   solution since each prime can be computed independently of all other
*   primes.
* AUTHOR: Blaise Barney 11/25/95 - adapted from version contributed by 
*   Richard Ng & Wong Sze Cheong during MHPCC Singapore Workshop (8/22/95).
* LAST REVISED: 04/12/05
****************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define LIMIT     2500000     /* Increase this to find more primes */
#define PRINT     100000      /* Print a line after this many numbers */


int isprime(int n) {
int i,squareroot;
if (n>10) {
   squareroot = (int) sqrt(n);
   for (i=3; i<=squareroot; i=i+2)
      if ((n%i)==0)
         return 0;
   return 1;
   }
/* Assume first four primes are counted elsewhere. Forget everything else */
else
   return 0;
}


int main(int argc, char *argv[])
{
int n,                           /* loop variables */
    pc,                          /* prime counter */
    foundone;                    /* most recent prime found */

printf("Starting. Numbers to be scanned= %d\n",LIMIT);

pc=4;     /* Assume the primes less than 10 (2,3,5,7) are counted here */

for (n=11; n<=LIMIT; n=n+2) {
   if (isprime(n)) {
      pc++;
      foundone = n;
      /***** Optional: print each prime as it is found 
      printf("%d\n",foundone);
      *****/
      }			
   if ( (n-1)%PRINT == 0 ) 
      printf("Numbers scanned= %d   Primes found= %d\n",n-1,pc);
   }
printf("Done. Largest prime is %d Total primes %d\n",foundone,pc);
} 

File: mpi/examples/ser_prime.f
C ************************************************************************
C FILE: ser_prime.f
C DESCRIPTION:
C   This program generates primes. The approach taken is a "brute force"
C   method which requires increasingly greater amounts of cpu as the problem
C   size increases. It should lend itself well to an embarrassingly parallel
C   solution since each prime can be computed independently of all other
C   primes.
C AUTHOR: Blaise Barney 11/25/95 - adapted from version contributed by 
C   Richard Ng & Wong Sze Cheong during MHPCC Singapore Workshop (8/22/95).
C LAST REVISED: 04/02/05
C ************************************************************************
C Explanation of constants and variables
C   LIMIT                  = Increase this to find more primes 
C   PRINT                  = Print a line after this many numbers
C   rank                   = task identifier 
C   pc                     = prime counter
C   foundone               = most recent prime found 

      program prime

      integer LIMIT, PRINT
      parameter(LIMIT=2500000)        
      parameter(PRINT=100000)

      integer n, pc, foundone
      logical result

      print *,'Starting. Numbers to be scanned=',LIMIT
C     Assume first four primes are counted here
      pc = 4
      do n=11, LIMIT, 2
         call isprime(n,result)
         if (result .eqv. .true.) then
            pc = pc + 1
            foundone = n
C           ***** Optional: print each prime as it is found
C           print *, foundone
C           *****
         endif
         if (mod(n-1,PRINT).eq.0) then
            print *,'Numbers scanned=',n-1,' Primes found= ',pc
         endif
      enddo
      print *,'Done. Largest prime is ',foundone,' Total primes ',pc
      end


      subroutine isprime (n,result)
      integer n
      logical result
      integer i, squareroot
      real*4 realn
      if (n .gt. 10) then
         realn = real(n)
         squareroot = int (sqrt(realn))
         do i=3,squareroot,2
           if (mod(n,i) .eq. 0) then
              result = .false.
              return
           endif
         enddo
         result = .true.
         return 
C Assume first four primes are counted elsewhere. Forget everything else
      else 
        result = .false.
        return
      endif
      end
